src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1441,23 +1431,33 @@ private MemberDeclarationSyntax ParseTypeDeclaration(SyntaxList<AttributeListSyn
 
 #nullable enable
 
-        private TypeDeclarationSyntax ParseClassOrStructOrInterfaceOrRoleDeclaration(SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers)
+        private TypeDeclarationSyntax ParseClassOrStructOrInterfaceOrExtensionDeclaration(SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers)
         {
             Debug.Assert(this.CurrentToken.Kind is SyntaxKind.ClassKeyword or SyntaxKind.StructKeyword or SyntaxKind.InterfaceKeyword ||
-                this.CurrentToken.ContextualKind is SyntaxKind.RecordKeyword or SyntaxKind.RoleKeyword or SyntaxKind.ExtensionKeyword);
+                this.CurrentToken.ContextualKind is SyntaxKind.RecordKeyword ||
+                (this.CurrentToken.Kind is SyntaxKind.ExplicitKeyword or SyntaxKind.ImplicitKeyword && this.PeekToken(1).ContextualKind == SyntaxKind.ExtensionKeyword));
 
             // "top-level" expressions and statements should never occur inside an asynchronous context
             Debug.Assert(!IsInAsync);
 
             var outerSaveTerm = _termState;
 
-            if (tryScanRecordStart(out var keyword, out var recordModifier))
+            SyntaxToken? firstKeyword;
+            SyntaxToken? secondKeyword;
+            SyntaxToken mainKeyword;
```

Comment:
maybe `typeKeyword` with a comment on it like `// class/struct/interface/record/extension`

Reply:
acutally, definitely comment.  looking below it's not clear what all three are for.

Reply:
do we need all 3 at the same time?  don't you only have at most one of these?

Reply:
We need to discriminate based on the kind of type (class/struct/interface/record/extension). Keeping track of which of the two keywords is the "main" keyword achieves that (without having to introduce many boolean flags or an enum).